File: android/app/src/main/java/com/cashu/me/ui/mints/MintDetailInfoLoader.kt
package com.cashu.me.ui.mints

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import com.cashu.me.Core.Wallet.userFacingWalletMessage
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.currentCoroutineContext
import kotlinx.coroutines.ensureActive

internal enum class MintConnectionState(val label: String) {
    NotChecked("Not checked"),
    Checking("Checking…"),
    Online("Online"),
    Offline("Unreachable"),
}

/** Used from the UI coroutine scope; only the latest request may publish state. */
internal class MintDetailInfoLoader<Info : Any> {
    var info by mutableStateOf<Info?>(null)
        private set
    var connection by mutableStateOf(MintConnectionState.NotChecked)
        private set
    var errorMessage by mutableStateOf<String?>(null)
        private set
    private var requestID = 0L

    suspend fun load(fetch: suspend () -> Info?) {
        currentCoroutineContext().ensureActive()
        val request = ++requestID
        connection = MintConnectionState.Checking
        // Keep the recovery explanation visible until a retry succeeds.
        try {
            val fetched = fetch()
            currentCoroutineContext().ensureActive()
            if (request != requestID) return
            if (fetched == null) {
                connection = MintConnectionState.Offline
                errorMessage = "The mint did not respond."
            } else {
                info = fetched
                errorMessage = null
                connection = MintConnectionState.Online
            }
        } catch (cancellation: CancellationException) {
            throw cancellation
        } catch (error: Exception) {
            currentCoroutineContext().ensureActive()
            if (request != requestID) return
            connection = MintConnectionState.Offline
            errorMessage = error.userFacingWalletMessage
        } finally {
            // Cancellation is not evidence that the mint is offline. Do not let
            // cleanup from a superseded request stop a newer loading indicator.
            if (request == requestID && connection == MintConnectionState.Checking) {
                connection = MintConnectionState.NotChecked
            }
        }
    }
}

File: ios/CashuWallet/Views/Mints/MintDetailInfoLoader.swift
import Foundation
import Observation

/// Owns the latest metadata request without tying cached information to reachability.
@MainActor
@Observable
final class MintDetailInfoLoader<Info> {
    enum Connection {
        case notChecked, checking, online, offline
    }

    private(set) var info: Info?
    private(set) var connection: Connection = .notChecked
    private(set) var errorMessage: String?
    private var requestID = UUID()

    var isLoading: Bool { connection == .checking }

    func load(fetch: () async throws -> Info?) async {
        guard !Task.isCancelled else { return }
        let request = UUID()
        requestID = request
        connection = .checking
        // Keep the recovery explanation visible until a retry succeeds.

        do {
            let fetched = try await fetch()
            try Task.checkCancellation()
            guard requestID == request else { return }
            guard let fetched else {
                connection = .offline
                errorMessage = "The mint did not respond."
                return
            }
            info = fetched
            errorMessage = nil
            connection = .online
        } catch {
            // An older request must not overwrite a newer success or loading state.
            guard requestID == request else { return }
            if Task.isCancelled || error is CancellationError
                || (error as? URLError)?.code == .cancelled {
                connection = .notChecked
            } else {
                connection = .offline
                errorMessage = error.userFacingWalletMessage
            }
        }
    }
}
